hibernate-reactive-core/src/test/java/org/hibernate/reactive/OrphanRemovalTest.java

```diff
@@ -0,0 +1,102 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import io.vertx.ext.unit.TestContext;
+import org.hibernate.cfg.Configuration;
+import org.junit.Test;
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.ManyToOne;
+import javax.persistence.OneToMany;
+import java.util.HashSet;
+import java.util.Set;
+
+import static javax.persistence.CascadeType.PERSIST;
+import static javax.persistence.CascadeType.REMOVE;
+
+
+public class OrphanRemovalTest extends BaseReactiveTest {
+
+    @Override
+    protected Configuration constructConfiguration() {
+        Configuration configuration = super.constructConfiguration();
+        configuration.addAnnotatedClass( Product.class );
+        configuration.addAnnotatedClass( Shop.class );
+        return configuration;
+    }
+
+    @Test
+    public void testOrphan(TestContext context) {
+        Shop shop = new Shop("shop");
+        shop.addProduct(new Product("ap1", shop));
+        shop.addProduct(new Product("ap2", shop));
+        shop.addProduct(new Product("ap3", shop));
+        shop.addProduct(new Product("ap4", shop));
+
+        test(context,
+                getSessionFactory()
+                        .withTransaction((session, transaction) -> session.persist(shop))
+                        .thenCompose(v -> getSessionFactory().withTransaction((session, transaction) -> session.find(Shop.class, shop.id)
+                                .thenCompose( result -> session.fetch(result.productList).thenApply(pl -> result) )
+                                .thenAccept(result -> {
+                                    // update
+                                    result.productList.clear();
+                                    result.addProduct(new Product("bp5", result));
+                                    result.addProduct(new Product("bp6", result));
+                                    result.addProduct(new Product("bp7", result));
+                                })
+                        ))
+                        .thenCompose(v -> getSessionFactory().withTransaction((session, transaction) ->
+                                session.createQuery("select count(*) from Product", Long.class)
+                                        .getSingleResult().thenAccept(result -> context.assertEquals(3l, result))
```

Comment:
Would it be too much to ask to check the content of the list and not just the size?

Reply:
I mean, the content of the table

Reply:
Well, if it's of length 3 instead of 4 or 7, I think we can be pretty certain orphan-delete worked.

Remember we're not trying to write tests for the whole of Hibernate core here.

Reply:
I don't agree with this but OK

Reply:
Well I guess I'm not sure what it is you want me to test?

That we didn't delete the wrong `Product`s? How could it possibly happen that we deleted the wrong four ids by accident, unless there was something seriously very broken in core?

Or something else?

Reply:
> How could it possibly happen that we deleted the wrong four ids by accident, unless there was something seriously very broken in core

This is actually why I write tests, make sure that somebody at some point doesn't mess up things.
We also don't always use the exact ORM code, or we might not use it in the future and having some more accurate tests  now makes me sleep better at night.

In this case in particular, it's not difficult to check the content of the list with an assertion after changing the `select count(*)` to a `select  name`.
Not sure why you are so against it, but I'm OK merging this if you feel so strong about it.

Reply:
I think we have to be able to trust in the Hibernate test suite (which is *massive*). If we start adding all sorts of tests for things that aren't implemented in Hibernate Reactive itself, we're going to quickly end up with a test suite that is as slow as the Hibernate test suite. I know that's not a problem for us *now*, since the tests run really fast *for now* but you would be surprised how quick and easy it is to bloat out a test suite and eventually find yourself always waiting for around for the build to finish. (This has happened to me over and over.)

Adding tests for all sorts of random things that are unlikely to break isn't smart. Smart is spending time thinking up things which *are* likely to break, and which *are* implemented with code in HR, and writing tests for those things.

Like here: instead of writing a more verbose test for the thing I could already see was working for sure, I asked myself "well, what happens to the associations belonging to the orphan itself?" and I wrote a test for that instead.

And guess what: buried treasure! My intuition lead me to something that was really quite broken.

Reply:
Agree to disagree.

And I don't find this test to be much more verbose:
```
session.createQuery("select name from Product", Long.class)
     .getResultList().thenAccept( list -> assertThat( list ).containsExactlyInAnyOrder( "bp5", "bp6", "bp7") )
```

But as I said before. Feel free to merge it.

Reply:
> And guess what: buried treasure! My intuition lead me to something that was really quite broken.

I cannot run your intuition on CI ... yet :)

Reply:
I don't know what `assertThat()` is. It's not a method of `TestContext`.

Reply:
```
import static org.assertj.core.api.Assertions.assertThat;
```

Reply:
`Assertions.assertThat` apparently.

Reply:
OK, I added that assertion.